### PR TITLE
8294193: Files.createDirectories throws FileAlreadyExistsException for a symbolic link whose target is an existing directory

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to fix the issue noted in https://bugs.openjdk.org/browse/JDK-8294193?

As discussed in a related PR https://github.com/openjdk/jdk/pull/10266#issuecomment-1246800456, this is a genuine bug in the `java.nio.file.Files.createDirectories` implementation. The commit in this PR fixes that by making sure that links are followed when checking if a `Path` is a directory during the creation of directories in the `createDirectories` implementation. The `private` method which is changed in this PR is only used in the `Files.createDirectories` and no other place, so this change should not impact any other code. Additionally a new test has been added to reproduce the issue and verify the fix.
tier1, tier2 and tier3 tests passed with this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294193](https://bugs.openjdk.org/browse/JDK-8294193): Files.createDirectories throws FileAlreadyExistsException for a symbolic link whose target is an existing directory

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10383/head:pull/10383` \
`$ git checkout pull/10383`

Update a local copy of the PR: \
`$ git checkout pull/10383` \
`$ git pull https://git.openjdk.org/jdk pull/10383/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10383`

View PR using the GUI difftool: \
`$ git pr show -t 10383`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10383.diff">https://git.openjdk.org/jdk/pull/10383.diff</a>

</details>

### CODE REVIEW

**Comment by DamonFool** [comment]
It passed on our machine with this patch.
Thanks for fixing it.

**Comment by AlanBateman** [comment]
I think create a new issue for Files.createDirectories issue or else move the existing issue from hotspot/runtime and change the description.

I plan to review this change but not today because it require digging into some security issues.

For the test then I think it will mean moving the existing tests from Misc.java to the new source file. That will fit with the description that you propose for the new test.

**Comment by jaikiran** [comment]
Hello Alan,

> I think create a new issue for Files.createDirectories issue or else move the existing issue from hotspot/runtime and change the description.

I've now created a separate JBS and this PR has been updated to link to that JBS.

> For the test then I think it will mean moving the existing tests from Misc.java to the new source file. That will fit with the description that you propose for the new test.

I've updated the PR to move the `Files.createDirectories` testing from `Misc.java` to this new test class. I have done some trivial cosmetic/non-functional changes to this moved test (things like printing an assert message if/when any assertion fails in that test).

> I plan to review this change but not today because it require digging into some security issues.

Please take your time.

File: test/jdk/java/nio/file/Files/CreateDirectories.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    @Test
+    public void testCreateDirectories() throws IOException {
+        final Path tmpdir = TestUtil.createTemporaryDirectory();
+        // a no-op
+        Files.createDirectories(tmpdir);
+
+        // create one directory
+        Path subdir = tmpdir.resolve("a");
+        Files.createDirectories(subdir);
+        assertTrue(Files.exists(subdir), subdir + " was expected to exist, but didn't");
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/CreateDirectories.java:100]: I realise this is moved from Misc.java but if you replace the "exists" with "isDirectory" then it improve the testing.

File: test/jdk/java/nio/file/Files/CreateDirectories.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    public void testSymlinkDir() throws Exception {
+        // create a temp dir as the "root" in which we will run our tests.
+        final Path startingDir = Files.createTempDirectory(Path.of("."), "8293792");
+        if (!TestUtil.supportsLinks(startingDir)) {
+            System.out.println("Skipping tests since symbolic links isn't " +
+                    "supported under directory "+ startingDir);
+            throw new SkipException("Symbolic links not supported");
+        }
+        System.out.println("Running tests under directory " + startingDir.toAbsolutePath());
+        final Path fooDir = Files.createDirectory(Path.of(startingDir.toString(), "foo"));
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/CreateDirectories.java:56]: You can simplify this to `Files.createDirectory(top.resolve("foo"))`, same for barDir, and symlink.

File: test/jdk/java/nio/file/Files/CreateDirectories.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ * @bug 8032220 8293792
+ * @summary Test java.nio.file.Files.createDirectories method
+ * @library ..
+ * @run testng CreateDirectories
+ */
+public class CreateDirectories {
+
+    /**
+     * Creates a symlink which points to a directory that exists. Then calls Files.createDirectories
+     * method passing it the path to the symlink and verifies that no exception gets thrown
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/CreateDirectories.java:44]: I think you can simplify the wording with something like "Test Files.createDirectories symbolic file with an existing directory."

File: test/jdk/java/nio/file/Files/CreateDirectories.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    @Test
+    public void testSymlinkDir() throws Exception {
+        // create a temp dir as the "root" in which we will run our tests.
+        final Path startingDir = Files.createTempDirectory(Path.of("."), "8293792");
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/CreateDirectories.java:49]: The tests in this area use TestUtil.createTemporaryDirectory() so it's probably best to keep it consistent if you can. 

Also can you rename "startingDir" to "top" as that will make the test much easier to read.

**Comment by jaikiran** [comment]
Hello Alan, I've updated the PR to implement your suggestions in the test. The test continues to pass.

**Review by AlanBateman — APPROVED** [review]
Thanks for the updates and for taking this issue.

**Comment by jaikiran** [comment]
Thank you for the review, Alan.
